ibis/expr/analysis.py

```diff
@@ -217,8 +176,12 @@ def substitute(fn, expr):
             arg = substitute(fn, arg)
         new_args.append(arg)
 
-    new_node = type(node)(*new_args)
-    return new_node.to_expr()
+    try:
+        new_node = node.__class__(*new_args)
```

Comment:
This should never fail, swallowing the exception here can cause weird behaviours.

Reply:
Yes, but unfortunately there's code in pushdown and projection optimization that depends on the exception to indicate that substitution wasn't possible.